core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -158,17 +158,14 @@ class ZooKeeperClient(connectString: String,
           inReadLock(initializationLock) {
             send(request) { response =>
               responseQueue.add(response)
-              inFlightRequests.release()
-              countDownLatch.countDown()
             }
           }
-        } catch {
-          case e: Throwable =>
-            inFlightRequests.release()
-            throw e
+       } finally {
+          inFlightRequests.release()
+          countDownLatch.countDown()
```

Comment:
Hmm.. I'm not sure about this. The point of counting down in the response handler is to make `handleRequests` synchronous. I think we may have to fix this problem in the session handler. We probably have to find a solution which does not require blocking until the expiration event is handled.

cc @junrao You might have some insight about the existing logic.